src/qmapshack/helpers/CProgressDialog.cpp

```diff
@@ -30,14 +30,16 @@ QStack<CProgressDialog*> CProgressDialog::stackSelf;
 CProgressDialog::CProgressDialog(const QString text, int min, int max, QWidget *parent)
     : QDialog(parent)
 {
-    if(!stackSelf.isEmpty())
```

Comment:
@kiozen what is the neccessity of this stack? When removing it displaying the progress works fine except the minor drawback (not sure if it really is a drawback) that there are progresses appearing above each other.

Reply:
The stack was introduced as everything else irritated users.

Reply:
I'm not sure wh it irritates users if there is the optimization progress bar and sometimes a routing progress appears above it. I must say I prefer if a program shows a bit too much that it is doing something. I understand that you don't want to have 100 progresses above each other on the screen, but since closing one screen will reactivate the underlyng one it doesn't change anything for me.

What I think the reason is that the bar is not properly shown is that a new bar immediately disables the old upper one but only appears one second later. So if there are routing progress bars appearing every 0.5s no progress bar will be shown.

Do you know if Progress bars are created from different threads and/or on the heap so construction order is not destruction order?

Reply:
In the beginning I used QProgressBar and we had several progress bars on the screen. On some window managers the blocked ones where overlapping the active ones blocking the active cancel buttons. I guess that is the reason for the stack. If I recall right :)

The progress bar is a design flaw. Qt has a cooperative multitasking. For GUI related stuff there can be only one thread. The main thread. The rule is to have no method that takes very long. By that the event loop is operated frequently and everything works nicely in a serialized way without blocking. 

For lengthy operations you are in trouble. Either you execute them in a thread. In that case you have to use events or scheduled signals to trigger feedback in the GUI. Additionally you have to take much care to serialize data access.

Or you block Qt. In that case you need the feedback of the progress bar. But to show the progress dialog and to process mouse events (click cancel) you need to process the event loop. And that is the moment where shit hits the fan. If the dialog would have it's own eco system everything would be fine. But it relies on the main event loop. Thus calling the event loop introduces re-entrancy. Something you do not take into account when writing software with Qt as everything is serialized. 

Now if you have several progress dialogs that are not independent, e.g. overall routing progress is interrupted by finding alternatives cancelling the overall progress dialog does not necessarily cancel the other ones. Therefore showing only one is the way to go.

This complete progress bar system is crap. But I did not find a better solution.

Reply:
I now moved the pausing of the old top progress bar to the moment the new one appears and the optimization one stills shows nicely and we only have one bar at once.

I changed the modality of the progress bar to application modal instead of window modal, as I don't want any user input on the element except of the cancel button. Are there use-cases in which the modality is only to stay on-top (window modal with nullptr parent) or the where user input shpuld only be blocked only for the parent and grandparent windows (window modal with valid parent)? I assume you can't exclude it, but if you know that it is the case then I'll find a solution in which the behaviour of the old code stays the same regarding modality instead of looking through the code to check every single use :wink:

Reply:
Oh, good catch. If the progress bar dialog is visible no other input except the cancel button should be possible.